internal/fans/nvidia.go

```diff
@@ -0,0 +1,228 @@
+package fans
+
+import (
+	"errors"
+	"os"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+	"github.com/markusressel/fan2go/internal/configuration"
+	"github.com/markusressel/fan2go/internal/ui"
+)
+
+type NvidiaFan struct {
+	Label             string                  `json:"label"`
+	Index             int                     `json:"index"`
+	RpmMovingAvg      float64                 `json:"rpmMovingAvg"`
+	Config            configuration.FanConfig `json:"config"`
+	MinPwm            *int                    `json:"minPwm"`
+	StartPwm          *int                    `json:"startPwm"`
+	MaxPwm            *int                    `json:"maxPwm"`
+	FanCurveData      *map[int]float64        `json:"fanCurveData"`
+	Rpm               int                     `json:"rpm"`
+	Pwm               int                     `json:"pwm"`
+	RunAtMaxSpeed     bool                    // to emulate PWM mode 0
+	SupportedFeatures int                     // (1 << FeaturePwmSensor) | (1 << FeatureRpmSensor) or similar
+	// TODO: probably SupportedFeatures doesn't have to be saved to config? set in GetDevices()
+	// TODO: put nvml.Device here? though in reality it should be shared
```

Comment:
Is it _required_ to share the `nvml.Device` instance or is it ok to have multiple instances?
I don't have a good way yet to deal with shared things like that, so if it works with multiple instances, I think this would be fine for a first implementation.

Reply:
You could have multiple instances, I think. Or probably you always get the same one anyway..

It's like this: One must call `nvml.Init()` before iterating the devices, and should call `nvml.Shutdown()` when the application shuts down.
This is ideally done centrally and only once.

And then you can iterate the device and see which one matches your ID in the config (e.g. "nvidia-10de2489-0400") best and get that handle.

But:
1. When / from where should `nvml.Init()` ideally be called? When `nvml.Shutdown()`?
2. When should the device handled be fetched? The `fans.Fan` and `sensors.Sensor` interfaces don't have an Init() function

Reply:
I have the rough idea of having some kind of Singleton in internal/nvidia/nvidia.go that has a function `GetDevice(id string) nvml.Device`, that everyone can call to get the handle to that device. And when it's called first, it will also call nvml.Init() - and when it has done that, it will use `runtime.SetCleanup()` for itself (that singleton object) which can then call `nvml.Shutdown()`.

Reply:
Sounds good, I didn't know about `runtime.SetCleanup()` :+1:

Reply:
damn, adding that singleton in `internal/nvidia/nvidia.go` introduces a cyclic dependency: with this, `internal/fans/nvidia.go` and `internal/sensors/nvidia.go` depend on `internal/nvidia/nvidia.go` (so they can get the nvml.device), but `internal/nvidia/nvidia.go` also depends on on the other two for `func GetDevices() []*NvidiaController`

Will I have to introduce yet another package (that they both depend on) to resolve this, or is there a better way?

Reply:
Put it into an extra package for now.

New problem: Neither `AddCleanup()` nor `SetFinalizer()` actually work with a global variable (even if it's allocated on the Heap, with `new`).
Simple solution: Call a cleanup function (that calls `nvml.Shutdown()` if it was initialized earlier) at the end of `main()`?

Reply:
> Simple solution: Call a cleanup function (that calls nvml.Shutdown() if it was initialized earlier) at the end of main()?

This is what I did for now